### ISSUE TITLE
Add JSON format

### ISSUE THREAD

**Issue by uce** [MaterializeInc/materialize#7186]
JSON is documented as a format, but it's only a data type.

We should consider adding a proper format for it, e.g.

```sql
 CREATE SOURCE s FROM FILE '...' FORMAT JSON;
```

Using format `JSON` could allow users to skip manually using format `BYTES` and casting to `jsonb`.

**Comment by aljoscha** [comment]
Thinking about it some more: I think it will be hard to determine the schema from a JSON input. With JSONB everything is dynamic, which comes with its own set of problems, thought.

**Comment by uce** [comment]
Yes, I think the most we could provide is to directly expose the data as `jsonb` instead of requiring the cast (see description).

**Comment by benesch** [comment]
From discussion with @morsapaes just now, we could follow [Flink's lead here](https://nightlies.apache.org/flink/flink-docs-master/docs/connectors/table/formats/json/) and ask that users tell us about a table type that matches their JSON schema. Here's how we'd want to do that in Materialize:

```sql
CREATE TYPE myrecord AS (a int, b text);
CREATE SOURCE ... FORMAT JSON MATCHING myrecord;
```

where the exact syntax of the `MATCHING myrecord` clause is up for debate. But the point is users would pre-declare their record type, and then tell us to assume that their JSON data complies with that schema. We could go one step further and introduce some inline sugar, too:

```sql
CREATE SOURCE ... FORMAT JSON MATCHING (a int, b text);
```

**Comment by benesch** [comment]
I'm really not sold on the `MATCHING` verb but I couldn't think of anything better in the moment.

**Comment by nmeagan11** [comment]
I prefer <code>USING SCHEMA</code> over <code>MATCHING</code>. It's described [here](https://github.com/MaterializeInc/materialize/issues/2679#issuecomment-614922357) but I also arrived at that conclusion independently. Having the inline sugar as @benesch outlines above would be ideal IMO, but I assume we'll have to implement <code>CREATE TYPE</code> regardless, meaning https://github.com/MaterializeInc/materialize/issues/3516 is a blocker on the SQL side (cc: @andrioni).

**Comment by andrioni** [comment]
A couple of points:
1. whatever we build here, IMO should also ideally work for all sources that can take JSON (Kafka, local file, S3)
2. we also need to add support for fetching JSON schemas from Confluent Schema Registry
3. we might also want to support reading JSON schemas (prior art: Redshift uses [JSONPath](https://goessner.net/articles/JsonPath/), Confluent uses [JSON Schema](https://json-schema.org/))
4. schemas shouldn't be mandatory, i.e. it should be easy to create something that just returns a single JSONB column instead (this is common in databases that have native support for JSON types, like Snowflake; Flink can't handle JSON gracefully with the Table API)
5. the behavior for casts (including failed casts) needs to be well specified (and potentially configurable)
6. the behavior for missing fields also needs to be well specified (and potentially configurable)
7. the behavior for handling bad messages (invalid UTF-8 strings, improper JSON, etc) also needs to be well specified (and potentially configurable)
7. we need to define exactly what format we expect and how it relates to each source (e.g. can you receive multiple JSON "rows" in a single Kafka message? Are they separated by `\n`? And how does it work for files? Do we support escaping/newlines inside strings?)

**Comment by benesch** [comment]
> 1. whatever we build here, IMO should also ideally work for all sources that can take JSON (Kafka, local file, S3)
> 5. the behavior for casts (including failed casts) needs to be well specified (and potentially configurable)
> 6. the behavior for missing fields also needs to be well specified (and potentially configurable)
> 7. the behavior for handling bad messages (invalid UTF-8 strings, improper JSON, etc) also needs to be well specified (and potentially configurable)
> 8. we need to define exactly what format we expect and how it relates to each source (e.g. can you receive multiple JSON "rows" in a single Kafka message? Are they separated by `\n`? And how does it work for files? Do we support escaping/newlines inside strings?)

Ugh, yeah, all of these fall under the “source orthogonality” umbrella. Protobuf in non-delimited sources doesn’t have an answer to (8), for example, and so we’re stuck disallowing it. All good questions to ask, but questions that are broader than JSON, and probably don’t need to block support for JSON over Kafka with the CSR.

> 2. we also need to add support for fetching JSON schemas from Confluent Schema Registry

We should make an issue for this. (Sorry, I would, but I’m on mobile and it’s late.)

> 3. we might also want to support reading JSON schemas (prior art: Redshift uses [JSONPath](https://goessner.net/articles/JsonPath/), Confluent uses [JSON Schema](https://json-schema.org/))

Yes! Prior issue on JSON schema: https://github.com/MaterializeInc/materialize/issues/1494

> 4. schemas shouldn't be mandatory, i.e. it should be easy to create something that just returns a single JSONB column instead (this is common in databases that have native support for JSON types, like Snowflake; Flink can't handle JSON gracefully with the Table API)

Like, a bare `FORMAT JSON` that’s identical to `FORMAT TEXT` except we cast to jsonb automatically for you?

**Comment by benesch** [comment]
> I prefer `USING SCHEMA` over `MATCHING`. It's described [here](https://github.com/MaterializeInc/materialize/issues/2679#issuecomment-614922357) but I also arrived at that conclusion independently. Having the inline sugar as @benesch outlines above would be ideal IMO, but I assume we'll have to implement `CREATE TYPE` regardless, meaning #3516 is a blocker on the SQL side (cc: @andrioni).

So that example you link is actually for a world where we accept a JSON schema in the string, rather than a SQL type. It occurs to me that `USING TYPE` is probably what we want to say! 

```sql
FORMAT JSON USING SCHEMA ‘some json schema’
FORMAT JSON USING TYPE foo
FORMAT JSON USING TYPE (int inline, int type, int def)
```

**Comment by andrioni** [comment]
>and probably don’t need to block support for JSON over Kafka with the CSR.
Agreed!

>Like, a bare FORMAT JSON that’s identical to FORMAT TEXT except we cast to jsonb automatically for you?
Yeah! For Snowflake, I think that's the extent of their support for loading JSON data: if you want a proper table with columns/etc, you just write the SQL manually accessing the fields from the JSON-typed column, no extra sugar on top.

**Comment by umanwizard** [comment]
There is another issue that has not been discussed here which is what to do in cases when the format is not naturally delimited. E.g., what does a JSON-formatted file source look like? Reasonable answers include (in no particular order):

* It doesn't matter, because file sources don't exist in cloud
* Each line is a separate JSON value (less flexible because it doesn't allow arbitrary whitespace in the values)
* The file is JSON values concatenated one after the other (this is more flexible than the per-line solution, but doesn't allow decoding in parallel)
* The entire file is one JSON array, and the individual rows are the elements of that array.

**Comment by umanwizard** [comment]
I was wrong about "hasn't been discussed here"; it looks like Alessandro did mention it:

> we need to define exactly what format we expect and how it relates to each source (e.g. can you receive multiple JSON "rows" in a single Kafka message? Are they separated by \n? And how does it work for files? Do we support escaping/newlines inside strings?)

**Comment by pH14** [comment]
I've been reading through our history of JSON-format-related issues, and am trying to synthesize them here into an overall proposal. It seems like there's both room for discussion on both how each JSON document is modeled in Materialize, as well as how each source interacts with the format. For the moment I'm only going to focus on the former issue, since it seems like it needs to be answered first and can be solved independently.

To model each document, it seems like a few options have been floated:

## Option 0: `CREATE SOURCE s ... FORMAT BINARY`

[Our existing JSON ingestion today](https://materialize.com/docs/sql/create-source/json-kafka/) takes in docs as binary, and requires users to manually convert/cast their data into `jsonb`. It's not the most ergonomic to work with, and makes JSON harder to work with than formats like Avro.

## Option 1: `CREATE SOURCE s ... FORMAT JSON`

This option provides a direct JSON format without any further schema information. This is similar to Option 0, except we do the conversion into a single `jsonb` value automatically.

Given an example document:

```json
{
	"a": 10,
	"b_outer": {
		"b_inner": 20,
	}
}
```

Access might then look like:

```sql
SELECT (data->a)::int, (data->b_outer->b_inner)::int FROM s;
```

Any views built on top of this source would need to annotate columns with casts.

I imagine adding this in would be fairly straightforward, and in some cases could provide an immediate ergonomic benefit over our current JSON solution. Having a single value JSON value to manipulate is similar to what Snowflake currently does.

The downsides of this approach are the continued lack of real type information, and having to manipulate all data via JSON functions. Over time and with complex data models, this could be painful and doesn't fit with how people interact with data encoded in other formats.

## Option 2: `CREATE SOURCE s ... FORMAT JSON USING TYPE (...)`

Our next option, as noted by @benesch and @nmeagan11, is to implement named composite types to represent the JSON document. 

When a source is created, it could specify the document structure through a named type:

```sql
CREATE TYPE bar (b_inner int);
CREATE TYPE foo (a int, b_outer bar);

CREATE SOURCE s ... FORMAT JSON USING TYPE foo;
```

Or inlined:

```sql
CREATE SOURCE s ... FORMAT JSON USING TYPE (a int, b_outer ROW(b_inner int));
```

Accessing these values becomes:

```sql
SELECT a, (b_outer).b_inner FROM s;
```

This approach has the pre-req of actually [implementing named composite types](https://github.com/MaterializeInc/materialize/issues/3516), which has the added benefit of getting us closer [to supporting `json_populate_record`](https://github.com/MaterializeInc/materialize/issues/2679). Modeling the document structure entirely through SQL types seems powerful, and this approach leaves little ambiguity about how each field maps to a column.

Among the downsides is that the Postgres spec specifically does not allow constraints like `NOT NULL` in composite type definitions, so we don't have an easy way to specify per-field nullability information or other constraints. Converting large JSON definitions into composite types could also be tedious and error prone to read and write.

## Option 3: `CREATE SOURCE s ... FORMAT JSON USING (CONFLUENT SCHEMA REGISTRY | SCHEMA = ' ' ) ...`

Our next option is to use an existing standard, like [JSON Schema](https://json-schema.org/) as the schema. From the schema, we can generate the right columns and validate each incoming JSON document against it. JSON Schema gives us richer information than straight composite types, like which fields are required and constraints over individual fields.

The downsides of this approach include: many (most?) sources may not have existing JSON Schemas, and JSON Schema handling is non-trivial and I'm not sure how mature the existing libraries are.

---

I hope that's an accurate / fair understanding of what folks have mentioned so far. From my read on the options, it seems like:

* Given that we're having any conversation here, Option 0 is insufficient. 
* Option 1 is mostly the same as Option 0, but might be nice to have. My $0.02 is that it does not substantially move the needle on Materialize's JSON support, so it probably doesn't take priority. (Maybe it would be useful to add _after_ we have other more robust JSON support, so we have a really quick way people can get JSON data into Materialize to play around with?)
* Option 2 seems promising, and it seems useful to have named composite types regardless of whatever else we do here.
* Option 3 is valuable down the line, and in implementation could build off the work in Option 2 (e.g. having a JSON Schema could auto-generate named composite types)

So then it might be reasonable to tackle this problem in order of: add in named composite types --> add in `FORMAT JSON USING TYPE ...` support --> add in JSON Schema / schema registry support. Independently, we can start ironing out how JSON will interact with individual sources, like delineation in files and Kafka messages, and join that effort up with the implementation once we have named composite types.

Open questions:
* Is anything I just wrote actually accurate? reasonable? 😄 
* Do users need a way to flatten nested fields? To do so with any of the approaches here, a user would need to define a new view that flattens access to specific fields.
* Is being able to specify required/nullable fields a requirement?

**Comment by benesch** [comment]
> Is anything I just wrote actually accurate? reasonable? 

Heh, I'd meant to DM you today to offer to share the context I had, but you pieced together the context 100% accurately without me. 🙌🏽 

> Among the downsides is that the Postgres spec specifically does not allow constraints like NOT NULL in composite type definitions, so we don't have an easy way to specify per-field nullability information or other constraints.

We actually _do_ support `NOT NULL` constraints in composite types! @umanwizard added support for it a while back to improve the typing of Avro records.

I'm still a bit scared there are weird edge cases in our PostgreSQL compatibility story here, but so far it's been working great.

> Converting large JSON definitions into composite types could also be tedious and error prone to read and write.

Definitely! But what else are you going to do? Unless you've got a JSON schema lying around, you've got to write down the schema _somehow_, and doing it via our native SQL types seems pretty solid. We could pivot the `json-flattened-view-gen` tool (https://github.com/MaterializeInc/materialize/pull/9943) to generate the requisite composite types from a sample datum to at least give users a starting point for the sequence of `CREATE TYPE` commands that they'll need to write.

> Do users need a way to flatten nested fields? To do so with any of the approaches here, a user would need to define a new view that flattens access to specific fields.

This is a good question. `FORMAT AVRO` always flattens the top-level record for historical reasons, but this is kinda bad because you can't opt-out of it. I think at some point @umanwizard mentioned that really that behavior ought to be opt-in via `ENVELOPE FLATTEN`. That could be something we could try with the new JSON support. 

----

Your proposed order of operations sounds awesome. 1b will be gnarly, and I don't have much to offer on that front.

**Comment by umanwizard** [comment]
I think it’s fine to unroll the top-level record into columns since that’s what we do in every other format and it seems to be what users expect/want. A record is roughly equivalent to a RelationType anyway, so it makes sense for the outer record to correspond to the outer type, i.e. the top-level columns.

**Comment by pH14** [comment]
Named composites have landed and https://github.com/MaterializeInc/materialize/pull/10977 is the in-progress PR for the first stab at JSON sources. The PR implements Option 2 only, so `FORMAT JSON USING TYPE t`. (Option 1 can be added in a follow-up PR and should be pretty quick).

We'll have to make some decisions on how to initially handle some data quality edge cases. After [some discussion on Slack](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1646405389947099), it seems like the default behavior should be strict, and leave room for finer-grained control down the line. Here's how that philosophy maps on to the scenarios I've seen so far:

### Missing Fields

**Scenario**: an JSON record is missing a field that is expected of its custom type. e.g. `CREATE SOURCE ... FORMAT JSON USING TYPE (a int, b text)` gets an input `{"a": 1}` with no `b` value.

**Current behavior**: throw an error

**Immediate workaround(s)**: the source should update its JSON output to use an explicit `null` value to imply an absent value, rather than omitting the field, e.g. `{"a": 1, "b": null}`

**Possible future featurework**: we enrich `CREATE TYPE` to support constraints like `NULL` and `NOT NULL` and default values, so users can more clearly control what should happen if a field is missing 

### Type mismatches

**Scenario**: a JSON value has the wrong type, e.g. `"true"` instead of `true` for a boolean field

**Current behavior**: do not attempt to coerce types, throw an error

**Immediate workaround(s)**: depending on what types the field can take on, use a `jsonb` column to allow the field to take on different shapes 

**Possible future featurework**: maaayybeeee add some type coercions, like strings to booleans or numbers

### Invalid JSON

**Scenario**: a JSON record is malformed and does not parse (e.g. `{"a": 1`)

**Current behavior**: throw an error

**Immediate workaround(s)**: fix the upstream source

**Possible featurework**: none

### Handling explicit `null` values

**Scenario**: a JSON record sets a field to `null`

**Current behavior**: Allow any field to be `null` if its JSON value is `null`. This follows the JSON spec which says `null` is always a valid value.

This isn't necessarily an issue, just want to state it out loud in case letting any field be `null` sets off alarm bells for anyone.

cc @nmeagan11 @andrioni do the behaviors mentioned above seem like a reasonable starting place?

**Comment by benesch** [comment]
Oh, d’oh, it occurs to me there’s prior art for this in PostgreSQL. It’s called [`jsonb_populate_record`](https://www.postgresql.org/docs/current/functions-json.html). Probably mentioned somewhere above or in one of the linked issues. 

@pH14, I haven’t looked at your code yet, but how hard would it be to pivot it to implement `jsonb_populate_record`, and then just implement `FORMAT JSON USING TYPE <typ>` as essentially syntax sugar for `SELECT jsonb_populate_record(NULL::<typ>, datum) FROM src (datum)`? (It’s not _quite_ syntax sugar, but I’d imagine there’s a way to slice it so you share almost 100% of the code between the source decoding pipeline and the impl of `jsonb_populate_record`.)

**Comment by pH14** [comment]
> @pH14, I haven’t looked at your code yet, but how hard would it be to pivot it to implement jsonb_populate_record, and then just implement FORMAT JSON USING TYPE <typ> as essentially syntax sugar for SELECT jsonb_populate_record(NULL::<typ>, datum) FROM src (datum)? (It’s not quite syntax sugar, but I’d imagine there’s a way to slice it so you share almost 100% of the code between the source decoding pipeline and the impl of jsonb_populate_record.)

Don't think that'd be too bad at all. I had actually been thinking of going the other way, and getting the first cut of a JSON source out first, and then reusing the code to implement `jsonb_populate_record` afterwards. Mostly started in this direction since implementing a new format is more unfamiliar to me than new SQL function at the moment, and thought it might be easier to iron out the unknowns there first

**Comment by benesch** [comment]
Yeah, I was thinking of it mostly as an easy way to defer questions about how to handle missing keys, NULLs, etc to Postgres!

**Comment by benesch** [comment]
Oh, and also, we can mark `FORMAT JSON USING TYPE` as experimental if you'd like to merge it while we're still thinking about the semantics!

**Comment by pH14** [comment]
> Yeah, I was thinking of it mostly as an easy way to defer questions about how to handle missing keys, NULLs, etc to Postgres!

Oooooh I see, yeah that's definitely intriguing. It looks like Postgres does a lot of type coercion (e.g. strings are coerced into the record's field types if possible) + allows anything to be null, so it's a much more permissive model.

> Oh, and also, we can mark FORMAT JSON USING TYPE as experimental if you'd like to merge it while we're still thinking about the semantics!

👍 is there anything special that needs to be done to mark it as experimental? I think this is a good call, since the current changes should be close to ready for devex et al to play around with

**Comment by benesch** [comment]
> > Yeah, I was thinking of it mostly as an easy way to defer questions about how to handle missing keys, NULLs, etc to Postgres!
> 
> Oooooh I see, yeah that's definitely intriguing. It looks like Postgres does a lot of type coercion (e.g. strings are coerced into the record's field types if possible) + allows anything to be null, so it's a much more permissive model.

Btw, be careful that you're not testing _literal_ coercion vs string conversion. `'1'::int` is allowed, but `'1'::text::int` is not. 

> > Oh, and also, we can mark FORMAT JSON USING TYPE as experimental if you'd like to merge it while we're still thinking about the semantics!
> 
> 👍 is there anything special that needs to be done to mark it as experimental? I think this is a good call, since the current changes should be close to ready for devex et al to play around with

Ah yeah, you'll just need to drop a call to `require_experimental_mode` to the path that plans `FORMAT JSON`. Here's an example:

https://github.com/MaterializeInc/materialize/blob/2c340fac61ee92ffb573e85d6735bf3c94a9a087/src/sql/src/plan/statement/ddl.rs#L239

**Comment by pH14** [comment]
Aright https://github.com/MaterializeInc/materialize/pull/11181 starts to go down the road of implementing `jsonb_populate_record`. I think it'll be a bit before it's ready, and the JSON source PR will be blocked until it's done.

The changes we'll need to get out are:
1. Finish + merge `jsonb_populate_record`
2. Update catalog reading so views can depend on custom types (currently an issue in the WIP JSON sources PR)
3. Then revise JSON source change to recycle logic from `jsonb_populate_record`
